HPCC-18905 Address EsdlConfig publish binding issues

rpastrana opened the pull request:
- Properly handle empty methodtree
- Improve empty string tests
- Query methods available based on new layout

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana commented:
@wangkx please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: ff64f736be7691196524834fb8af99564987f98b
Build: success
Install hpccsystems-platform-community_6.4.8-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |628 |628 |0 |
|test (roxie) |761 |761 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  158 sec (00:02:38) |   38 sec (00:00:38) |    5 sec (00:00:05) |   35 sec (00:00:35) | 1082 sec (00:18:02) |   32 sec (00:00:32) | 1355 sec (00:22:35) |

wangkx reviewed (approved):
@rpastrana looks good.

rpastrana commented:
@richardkchapman please merge (currently targeting 6.4.8, but can target 6.4.10 if the cutoff has already expired.)

richardkchapman commented:
I will wait to see whether there needs to be another rc

rpastrana committed:
HPCC-18905 Address EsdlConfig publish binding issues

- Properly handle empty methodtree
- Improve empty string tests
- Query methods available based on new layout
- Other onMethods utilize new layout

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

rpastrana commented:
@wangkx please re-review. I noticed other methods within this same webservice should be querying ESDL method info from EsdlService/EsdlMethod rather than just EsdlMethod. This is due to a change Yanrui put in.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 55b5047832b3c41db2fc046cc8f453d3a7a41498
Build: success
Install hpccsystems-platform-community_6.4.8-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |97 |97 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |628 |628 |0 |
|test (roxie) |761 |761 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  159 sec (00:02:39) |   37 sec (00:00:37) |    6 sec (00:00:06) |   36 sec (00:00:36) | 1072 sec (00:17:52) |   33 sec (00:00:33) | 1348 sec (00:22:28) |

wangkx reviewed (approved):
@rpastrana looks fine.

rpastrana commented:
@richardkchapman please merge
